File: examples/05-fork-join/Example7.java
// =================================================================
//
// File: Example7.java
// Authors:
//				 A01651517 Peter Glez
//				 A01703947 Juan Alcántara
// Description: This file contains the code to brute-force all
//				prime numbers less than MAXIMUM using Java's
//				Fork-Join.
//
// Copyright (c) 2021 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

import java.util.Arrays;
import java.util.concurrent.RecursiveAction;
import java.util.concurrent.ForkJoinPool;

public class Example7 extends RecursiveAction {
	private static final int SIZE = 1_000_000;
	private static final int MIN = 1_000;
	private int start, end;
	private boolean array[];

	public Example7 (boolean array[], int start, int end) {
		this.array = array;
		this.start = start;
		this.end = end;
	}

	void computeDirectly() {
		for (int i = start; i < end; i++) {
			array[i] = true;
			for (int j = 2; j <= Math.sqrt(i); j++) {
				if (i % j == 0) {
					array[i] = false;
					break;
				}
			}
		}
	}

	@Override
	protected void compute() {
		if ((start - end) <= MIN) {
			computeDirectly();
		} else {
			int mid = (start + end)/2;
			invokeAll(
					new Example7(array, start, mid),
					new Example7(array, mid, end) 
			);
		}
	}

	public static void main(String args[]) {
		long startTime, stopTime;
		boolean array[];
		double ms;
		ForkJoinPool pool;

		array = new boolean[SIZE];

		System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
		ms = 0;
		for (int i = 1; i <= Utils.N; i++) {

			pool = new ForkJoinPool(Utils.MAXTHREADS);

			startTime = System.currentTimeMillis();

			pool.invoke(new Example7(array, 2, array.length));

			stopTime = System.currentTimeMillis();
			ms += (stopTime - startTime);
			if (i == Utils.N) 
				for (int j = 2; j < 100; j++)
					if (array[j])
						System.out.printf("%d ", j);
		}
		System.out.printf("\navg time = %.5f ms\n", (ms / Utils.N));
	}

}

File: examples/06-openmp/example7.c
// =================================================================
//
// File: example7.c
// Author(s):
//					A01651517 Pedro Luis González Roa
//					A01703947 Juan Alejandro Alcántara Minaya
// Description: This file contains the code to brute-force all
//				prime numbers less than MAXIMUM using OpenMP.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "utils.h"

#define MAXIMUM 100000000 //1e6

void is_prime(int* a, int x) {
	if (a[x] != 0) { // Check if it has not been already set as no prime
		a[x] = 1; // Define as prime until proven wrong

    // Calculate if prime
		for (int i = 2; i < sqrt((double)x); i++)
			if (x % i == 0)
				a[x] = 0;

    // If prime then all the multiples are not prime
		if (a[x] == 1)
			for (int i = 2; x * i < N; i++)
				a[x * i] = 0;
	}
}

int main(int argc, char* argv[]) {
	int i, *a;
	double ms;

	a = (int *) malloc(sizeof(int) * (MAXIMUM + 1));
	printf("At first, neither is a prime. We will display to TOP_VALUE:\n");
	for (i = 2; i < TOP_VALUE; i++) {
		a[i] = -1;
		if (a[i] == -1) {
			printf("%i ", i);
		}
	}
	printf("\n");

	printf("Starting...\n");
	ms = 0;
	for (i = 0; i < N; i++) {
		start_timer();

    int j;
    #pragma omp parallel for shared(a)
    for (j = 0; j < MAXIMUM; j++)
      is_prime(a, j);

		ms += stop_timer();
	}
	printf("Expanding the numbers that are prime to TOP_VALUE:\n");
	for (i = 2; i < TOP_VALUE; i++) {
		if (a[i] == 1) {
			printf("%i ", i);
		}
	}
	printf("\n");
	printf("avg time = %.5lf ms\n", (ms / N));

	free(a);
	return 0;
}
